Custom select box styling

aaronshekey opened the pull request:
A stab at custom styling for our select dropdowns. They've always looked a little out of place on pages like this...

### Context before
![context](https://cloud.githubusercontent.com/assets/1369864/8067637/a3ff7fc4-0eb4-11e5-9fd0-55d986d8192a.png)

### Context after
![screen shot 2015-06-09 at 2 42 52 pm](https://cloud.githubusercontent.com/assets/1369864/8067835/e9ecf088-0eb5-11e5-8203-2b1153161828.png)

### Safari 8.0.6
![screen shot 2015-06-09 at 2 31 41 pm](https://cloud.githubusercontent.com/assets/1369864/8067645/b3205ba4-0eb4-11e5-90c2-d6465c793454.png)
![screen shot 2015-06-09 at 2 31 47 pm](https://cloud.githubusercontent.com/assets/1369864/8067648/b640a8c0-0eb4-11e5-9667-2956424ea52f.png)

### Retina
![screen shot 2015-06-09 at 2 39 38 pm](https://cloud.githubusercontent.com/assets/1369864/8067767/6d095a70-0eb5-11e5-8c8e-1eef299640c2.png)


### IE11 Windows 7
![dropdown](https://cloud.githubusercontent.com/assets/1369864/8067680/85d0e054-0eac-11e5-96d6-f176ef5e9cee.PNG)
![dropdown2](https://cloud.githubusercontent.com/assets/1369864/8067682/8bd6a560-0eac-11e5-906a-33abb47c58b0.PNG)

cc: @mdo @jonrohan

jonrohan commented on scss/_forms.scss:
```diff
@@ -469,3 +470,35 @@ input::-webkit-inner-spin-button {
+  border-radius: 3px;
+  box-shadow: inset 0 -1px 2px rgba(0, 0, 0, 0.075);
+  -webkit-appearance: none;
+  -moz-appearance: none;
```
Definitely, we have autoprefixer for that

aaronshekey replied:
I'm pretty sure appearance isn't covered by autoprefixer. When I drop it down to just `appearance: none;` none of the styling happens.

mdo replied:
I added a comment to the code for why we need this—autoprefixer only prefixes properties that are in the spec. `appearance` is not.

jonrohan replied:
Looks like appearance was added https://github.com/postcss/autoprefixer/blob/9497b6452e21e7ab69974f34ed215bf568b24db4/CHANGELOG.md#52-dont-tread-on-me to 5.2 https://github.com/postcss/autoprefixer/commit/d3e5a76638952347ba755325f5eb515dbb0916d4#commitcomment-11402190

jonrohan commented on scss/_forms.scss:
```diff
@@ -469,3 +470,35 @@ input::-webkit-inner-spin-button {
+  padding: 7px 24px 7px 8px;
+  vertical-align: middle;
+  background-color: #fff;
+  background: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABIAAAAaCAMAAACNQ/wIAAAAYFBMVEUzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMS2aZBAAAAH3RSTlMAAQYNDg8VJDQ2N0JbdXZ4h4mKpL3Iycvb6vDx8vn+GLdSCAAAAHBJREFUeNq90TcSgDAMRFGRMwZMTnv/W+IBm6BxzSt/pUBGURKT73v+LdkGbNm7xDOUOXlKNOE0RaaEI7QxvIrX4db5pLgtXlqXyJH4kA5VYCr ...
```
We don't typically base64 images, but not saying it's a bad thing. Just asking the obvious if we're cool with this.

aaronshekey replied:
I'm fine doing this as a static asset someplace, but would need advice as to how we set that up.

mdo replied:
Let's leave it as base64 since Primer itself has no CDN.

jonrohan commented:
Overall I'm :+1: on this kind of improvement. I personally would like it to look slightly closer to our buttons than our text boxes, just to give it a little more dropdownish feel.

jonrohan commented:
![screen shot 2015-06-09 at 4 48 35 pm](https://cloud.githubusercontent.com/assets/54012/8069289/682c64f4-0ec7-11e5-9b06-265549f49321.png)

An afterthought, if we go down this road, should we make our other custom dropdowns match? Or at least use the double arrow icon?

aaronshekey commented:
I'm fine with there being a difference between true select boxes having the double arrow and the drop downs having the single. The former is a true form element where the latter is just a navigation item. I can add a stronger gradient to match the buttons, but I'll defer to @mdo on the direction we want to go with that. I like that these are currently the inverse of an input.

mdo commented:
I'm actually kind of okay with making it look a bit different than our buttons and custom `.select-menu` component. The behavior is a lot different between those two elements, so conflating the two could be harmful from a UX side.

Originally I wasn't into the bottom-up shadow, but it feels push-y enough and looks pretty good still. I think the arrow needs to be a little more to the left, but that's about it.

Lastly, we might want to add some basic styles to ensure proper vertical alignment of non-styled `<select>`s. For example, we're changing the height there—should we do that for all `<select>`s? Bootstrap does this, which triggers the fallback styles for Webkit browsers, but it's a consistent height and alignment which is more helpful I think.

aaronshekey commented:
I've now added fallback styling to the select element, which basically means I've moved most of the visual styling to the element itself save for the arrow overrides.

### Safari
![screen shot 2015-06-09 at 5 18 41 pm](https://cloud.githubusercontent.com/assets/1369864/8070985/9d14e11a-0ecb-11e5-875e-f09a0e531b60.png)

### Chrome
![screen shot 2015-06-09 at 5 20 56 pm](https://cloud.githubusercontent.com/assets/1369864/8071016/ee863558-0ecb-11e5-8eeb-1fd6e4408540.png)


Should be rad. `Select`s also take the same styling as their `input` counterparts in Form Groups
![screen shot 2015-06-09 at 5 19 45 pm](https://cloud.githubusercontent.com/assets/1369864/8070999/cae22170-0ecb-11e5-9c62-8efa34fd5911.png)

mdo commented:
Pushed a bunch of changes clean up a few things:

- Broke apart the default and custom selects to better explain what each does.
- Updated the default `<select>` styles to target any element that *doesn't* have `[multiple]`.
- Renamed `.select-small` to `.select-sm` to match our buttons.
- Fixed the broken form group styles example.
- Excluded the forms file from SCSS lint to avoid those vendor prefix warnings.

Here are some updated screenshots showing the latest:

![screen shot 2015-06-10 at 12 36 42 pm](https://cloud.githubusercontent.com/assets/98681/8092434/8b3ba67e-0f6e-11e5-8288-985f0c33440e.png)
![screen shot 2015-06-10 at 12 36 51 pm](https://cloud.githubusercontent.com/assets/98681/8092436/8b549f12-0f6e-11e5-8585-2e0acc56531c.png)
![screen shot 2015-06-10 at 12 36 59 pm](https://cloud.githubusercontent.com/assets/98681/8092435/8b5230ce-0f6e-11e5-81d3-c99a48371083.png)
![screen shot 2015-06-10 at 12 37 07 pm](https://cloud.githubusercontent.com/assets/98681/8092437/8b55afa6-0f6e-11e5-8a18-27e85aed4eea.png)

jonrohan commented on scss/_forms.scss:
```diff
@@ -469,3 +480,42 @@ input::-webkit-inner-spin-button {
+  // scss-lint:disable VendorPrefix
+  // Have to include vendor prefixes as the `appearance` property isn't part of the CSS spec.
+  -webkit-appearance: none;
+  -moz-appearance: none;
```
I guess houndci doesn't like it. :-/

jonrohan commented:
![image](https://cloud.githubusercontent.com/assets/54012/8092830/aa73b2cc-0f8a-11e5-9389-b9f6ce7eaa8a.png)

no idea what is wrong with houndci

aaronshekey commented:
Loving where this is at. @mdo, Thanks for jumping in and fixing the last bits.

Sidenote: Maybe we should consider switching to postcss https://github.com/nDmitry/grunt-postcss instead of the deprecated thing we've got going on now.

mdo commented:
> Sidenote: Maybe we should consider switching to postcss https://github.com/nDmitry/grunt-postcss instead of the deprecated thing we've got going on now.

What's deprecated?

aaronshekey commented:
Is this what we're using? https://github.com/nDmitry/grunt-autoprefixer It's been deprecated in favor of grunt-postcss.

mdo commented:
Oh, yeah, we can do that for this release.
